### PR TITLE
Utils and libraries: Fix #13371 - move umask operation earlier in AppInit()

### LINKED ISSUE #13371: -sysperms=false (default) doesn't appear to do anything
Despite -sysperms=false (default) docs saying files will be created with perms 077, none of my default-created datadirs appear to have any permissions aside from 0600. Maybe I'm misreading the docs but it appears umask(077) isnt doing anything.

### PR DESCRIPTION
**Pull request by n2yen**
When the option -sysperms=false (default) is enabled, the created datadirs doesn't take into account the default umask setting of '077'.
The fix moves the umask operation to just after parameters are parsed during AppInit()

### CODE REVIEW

**Comment by laanwj** [comment]
Or, if this is a big issue, maybe do it in both places?

**Comment by MarcoFalke** [comment]
Would be nice to know when this regression was introduced. If it never did anything, I guess it can be removed?

If it is not going to be removed, this must come with functional tests that verify the behavior.

**Comment by n2yen** [comment]
Appreciate the feedback, I will look into these later - please let me know if you have any other suggestion as well.  Thank you!

**Comment by n2yen** [comment]
Had a look at the bitcoin-qt as suggested - it looks like if there was a regression it could be because it potentially generates datadir on its own, without calling GetDefaultDataDir(), see comments in the commit above. Tested by running bitcoind, and bitcoin-qt and checking datadir. The existing unit tests, don't appear to rely on either of the above utility function calls - should I still go ahead and add unit tests for this anyway? thanks!

**Comment by ken2812221** [comment]
@n2yen You can add a python script into functional test, it allow you to run a node to test this.

**Comment by laanwj** [comment]
> If it is not going to be removed, this must come with functional tests that verify the behavior.

It definitely works, but not for the initial directory creation. It works for files created after that, and that's what it's intended for (block file data sharing, basically).

**Comment by n2yen** [comment]
@Iaanwj can you clarify if we should also leave the old the old code where it was also? i'm not sure, thank you for your feedback!

**Comment by MarcoFalke** [comment]
@n2yen I haven't looked, but according to the issue  #13371 the old code is doing nothing. Having useless code is useless, so please remove if it is useless.

**Comment by laanwj** [comment]
So to show the current code should not naively be removed (nor applied for `bitcoind` only), this affects wallet default permissions:
```bash
$ src/bitcoind -datadir=/tmp/wallet1 -regtest
$ stat /tmp/wallet1/regtest/wallets/wallet.dat 
Access: (0600/-rw-------)
```
After removing the `umask` setting code from `AppInitBasicSetup`:
```bash
$ src/bitcoind -datadir=/tmp/wallet2 -regtest
$ stat /tmp/wallet2/regtest/wallets/wallet.dat
Access: (0660/-rw-rw----)
```
(wallet is group-readable now)

**Comment by n2yen** [comment]
@Iaanwj ok, so we're in agreement, that the umask in AppInitBasicSetup() may not be early enough.  I agree that the umask cannot just be in bitcoind.cpp, but also in bitcoin-qt or some common code that can be called earlier... I will also be adding tests in the way suggested by @ken2812221.

**Comment by n2yen** [comment]
Updated commit with tests to verify permissions of regtest directory, and umask is set as early as sensible in both bitcoind and bitcoin-qt.

**Comment by n2yen** [comment]
Rebased on top of:
eb2f1bd276108e70aff0f582a407e9b702eb4dd1 - Merge #14189: qa: Fix silent merge conflict in wallet_importmulti...
On sept 10, 2018

File: test/functional/feature_nosysperms.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+from test_framework.test_framework import BitcoinTestFramework
+
+class NoSyspermsTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.setup_clean_chain = True
+        self.num_nodes = 1
+
+    def check_permissions(self, fullpathname):
+        self.log.info('{}  {}'.format(stat.filemode(os.stat(fullpathname).st_mode), fullpathname))
+        return (os.stat(fullpathname).st_mode & (stat.S_IRWXO|stat.S_IRWXG)) == 0
```
Comment (practicalswift) [inline test/functional/feature_nosysperms.py]:
```
2018-09-14 14:23:42 flake8(pr=13389): test/functional/feature_nosysperms.py:19:62: E227 missing whitespace around bitwise or shift operator
```

File: test/functional/feature_nosysperms.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    def check_dir_permissions(self, regtest_dirpath):
+        result = []
+        for path, dirs, files in os.walk(regtest_dirpath):
+            for filename in files:
+                result.append(self.check_permissions(os.path.join(path, filename)))
+        return result
+
+    def run_test(self):
+        #self.stop_node(0)
+        #self.start_node(0, ['-regtest'])
```
Comment (practicalswift) [inline test/functional/feature_nosysperms.py]: Remove these two comments :-)

File: test/functional/feature_nosysperms.py (hunk 1 above)
Comment (n2yen) [reply test/functional/feature_nosysperms.py]: Done.

File: test/functional/feature_nosysperms.py (hunk 2 above)
Comment (n2yen) [reply test/functional/feature_nosysperms.py]: Done. Thanks for the feedback!

**Comment by ken2812221** [comment]
Concept ACK. 
nit: You should skip the test on Windows.

**Comment by n2yen** [comment]
Hi @ken2812221, 
Thanks for your feedback. I've made the changes, to skip the test on Windows, however I don't have a windows machine to confirm it skips. Can someone try it? I've confirmed the tests will run on Linux.

**Comment by n2yen** [comment]
Thanks Ken2812221, yup, I'll look into it.

**Comment by practicalswift** [comment]
@n2yen Job restarted :-)

**Comment by n2yen** [comment]
@practicalswift thank you!

File: test/functional/feature_sysperms.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    def run_test(self):
+        self.log.info("Check for valid platform")
+        if platform.system() == 'Windows':
+            raise SkipTest("This test does not run on Windows.")
+
+        self.restart_node(0)
+        datadir = self.nodes[0].datadir
+
+        self.log.info("Running test without -sysperms option - assuming umask of '0o77'")
```
Comment (hebasto) [inline test/functional/feature_sysperms.py]: `'0o77'` - typo?

Comment (n2yen) [reply test/functional/feature_sysperms.py]: you're right, should be '077'. What did you make of -sysperm option, is it still needed?

Comment (hebasto) [reply test/functional/feature_sysperms.py]: > What did you make of -sysperm option, is it still needed?

My concerns about `-sysperms` are in the note here #15902.

Comment (n2yen) [reply test/functional/feature_sysperms.py]: does it makes sense to close this PR - perhaps replace it with simply removing the option? i suppose the tests might still be useful...

**Comment by MarcoFalke** [comment]
Please squash your commits according to https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#squashing-commits

**Comment by KaosGhostNINJA** [comment]
I believe  that it might not work on windows till I finish what I am doing.

**Comment by n2yen** [comment]
rebased!

**Comment by jonasschnelli** [comment]
utACK 09072f762ae3fe9a746ab447458e95a6fb646ed3

**Comment by laanwj** [comment]
I don't think my original comments have been addressed. This still doesn't allow setting sysperm in the configuration file, only on the command line.

It also complicates the initialization order.

I'm starting to think it would be better to remove the `sysperm` setting.

**Comment by jonasschnelli** [comment]
I underestimated the fact that one can no longer use the `sysperm` setting in the config file after this. I just retrieved my ACK.

**Comment by jonasschnelli** [comment]
@n2yen: what is the status of this PR? Can you have a look in addressing/commenting @laanwj [comment](https://github.com/bitcoin/bitcoin/pull/13389#issuecomment-539906114)?

**Comment by n2yen** [comment]
ok, will have a look.

**Comment by jonasschnelli** [comment]
Closing in favour of #17127 (which is probably superior).

**Pull request closed without merging**